src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -388,42 +419,79 @@ private static bool IsRoutePartNonEmpty(object routePart)
         private struct TemplateBindingContext
         {
             private readonly RouteValueDictionary _defaults;
-            private readonly RouteValueDictionary _acceptedValues;
+            private RouteValueDictionary _acceptedValues;
+            private readonly RouteValueDictionary _values;
 
-            public TemplateBindingContext(RouteValueDictionary defaults)
+            public TemplateBindingContext(RouteValueDictionary defaults, RouteValueDictionary values)
             {
                 _defaults = defaults;
 
-                _acceptedValues = new RouteValueDictionary();
+                // Perf: Lazily Initialize the AcceptedValues of the context with the new values 
+                // AcceptedValues will be initialized when any of the values has to be updated.
+                // For most of the common cases, AcceptedValues is same as the values.
+                // In such cases, this optimization saves one RouteValueDictionary allocation per GetValues call.
+                _acceptedValues = null;
+                _values = values;
             }
 
             public RouteValueDictionary AcceptedValues
             {
-                get { return _acceptedValues; }
+                get
+                {
+                    return _acceptedValues ?? _values;
+                }
             }
 
             public void Accept(string key, object value)
             {
+                EnsureAccceptedValues();
                 if (!_acceptedValues.ContainsKey(key))
                 {
                     _acceptedValues.Add(key, value);
                 }
+                else
+                {
+                    var existingValue = _acceptedValues[key];
+                    if (!ReferenceEquals(existingValue, value))
```

Comment:
I know, but I did it intentionally. This class is private to TemplateBinder and this method is called with either of two values (one that is existing in values or _defaults) So I took that liberty of comparing by reference :)